subiquity/server/controllers/install.py

```diff
@@ -380,6 +385,61 @@ async def run_curtin_step(name, stages, step_config, source=None):
                     step_config=self.rp_config(logs_dir, mp.p()),
                     source='cp:///cdrom',
                     )
+            await self.create_rp_boot_entry(context=context, rp=rp)
+
+    @with_context(description="creating boot entry for reset partition")
+    async def create_rp_boot_entry(self, context, rp):
+        fs_controller = self.app.controllers.Filesystem
+        if not fs_controller.reset_partition_only:
+            cp = await self.app.command_runner.run(
+                ['lsblk', '-n', '-o', 'UUID', rp.path],
+                capture=True)
+            uuid = cp.stdout.decode('ascii').strip()
+            conf = grub_reset_conf.replace("#UUID#", uuid)
+            conf = conf.replace("#PARTITION#", str(rp.number))
+            with open(self.tpath('etc/grub.d/99_reset'), 'w') as fp:
+                os.chmod(fp.fileno(), 0o755)
+                fp.write(conf)
+            await run_curtin_command(
+                self.app, context, "in-target", "-t", self.tpath(), "--",
+                "update-grub", private_mounts=False)
+        if self.app.opts.dry_run and not is_uefi_bootable():
+            # Can't even run efibootmgr in this case.
+            return
+        state = await self.app.package_installer.install_pkg('efibootmgr')
+        if state != PackageInstallState.DONE:
+            raise Exception()
+        efi_state_before = get_efibootmgr('/')
+        cmd = [
+            'efibootmgr', '--create',
+            '--loader', '\\EFI\\boot\\shimx64.efi',
+            '--disk', rp.device.path,
+            '--part', str(rp.number),
+            '--label', "Restore Ubuntu to factory state",
+            ]
+        await self.app.command_runner.run(cmd)
+        efi_state_after = get_efibootmgr('/')
+        new_bootnums = (
+            set(efi_state_after.entries) - set(efi_state_before.entries))
+        if len(new_bootnums) == 0:
+            return
+        new_bootnum = new_bootnums.pop()
+        new_entry = efi_state_after.entries[new_bootnum]
+        was_dup = False
+        for entry in efi_state_before.entries.values():
+            if entry.path == new_entry.path and entry.label == new_entry.label:
+                was_dup = True
+        if was_dup:
+            cmd = [
+                'efibootmgr', '--delete-bootnum',
```

Comment:
I'd like some unit tests on this dup-cleanup functionality.  I want to reduce the likelihood of the wrong boot entries getting deleted.